### PR TITLE
[Unity][Transform] Implement Relax function inlining

### PR DESCRIPTION
**Pull request by Lunderberg**
This commit adds the ability to inline one Relax function into another.  This is provided with two APIs, one which allows explicit specification of the functions to be inlined, and one which inlines
every private Relax function in an `IRModule`.  Function inlining with explicit replacements is intended for use by an end-user when building a model function (e.g. inlining of a rotary embedding), and the second is intended for use as part of a generic optimization pipeline.

### CODE REVIEW

**Comment by Lunderberg** [comment]
This PR is currently marked as a draft, as it relies on improvements implemented in https://github.com/apache/tvm/pull/16193.

File: src/relax/transform/inline_functions.cc [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    if (auto callee = op->op.as<GlobalVar>()) {
+      callee_to_caller_[callee.value()].insert(current_caller_.value());
+    }
+  }
+
+  Optional<GlobalVar> current_caller_ = NullOpt;
+  GlobalVarMap callee_to_caller_;
+};
+
+std::unordered_set<GlobalVar, ObjectPtrHash, ObjectPtrEqual> FindRecursiveFunctions(
```
Comment (slyubomirsky) [inline src/relax/transform/inline_functions.cc:76]: https://github.com/apache/tvm/blob/unity/src/relax/analysis/detect_recursion.cc Not sure if my analysis might be suitable for this purpose.

Comment (Lunderberg) [reply src/relax/transform/inline_functions.cc]: Ooh, thank you for pointing that out, and that's exactly what I was looking for.  Updated the PR to use `DetectRecursion`.

File: src/relax/transform/inline_functions.cc [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      // from the MatchCast inlined portion.
+      //
+      // Option 4: Define a VarBinding `param[i] = args[i]`, with
+      // CanonicalizeBindings to handle conversion of Var to
+      // DataflowVar, and tvm::relax::Bind to handle substitution of
+      // symbolic variables.  This would result in a well-formed Relax
+      // function, with no duplicate definitions of symbolic
+      // variables.
+      //
+      // This implementation uses Option 4.
```
Comment (slyubomirsky) [inline src/relax/transform/inline_functions.cc:127]: Very clever solution to avoid dealing with dataflow blocks. It's good that `CanonicalizeBindings` has become a single source of truth for dealing with these issues.

File: tests/python/relax/test_inline_functions.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_subroutine_with_symbolic_vars():
+    """Inlined subroutines should use the caller's symbolic variables
+
+    Before inlining, the subroutine and the caller have distinct
+    `tir::Var` for each symbolic variables.  After inlining, only the
+    caller's `tir::Var` symbolic variables should remain.
+    """
```
Comment (slyubomirsky) [inline tests/python/relax/test_inline_functions.py:173]: What logic in the pass handles this replacement? The use of CanonicalizeBindings per "option 4"?

File: tests/python/relax/test_inline_functions.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            if cond:
+                Out = Before.subroutine()
+            else:
+                Out = R.const(0, "int64")
+
+            return Out
+
+    with pytest.raises(Exception):
+        Before["main"].inline_functions({"subroutine": Before["subroutine"]})
+
```
Comment (slyubomirsky) [inline tests/python/relax/test_inline_functions.py:360]: A couple of cases that are also worth testing:
1. Inlining within an inlined function
2. Making sure that mutually recursive functions can't be inlined either (since if you can inline within an inlined function, mutual recursion would also cause infinite inlining)

**Review by slyubomirsky — CHANGES_REQUESTED** [review]
Overall, this change looks great, and it is a long-awaited one. My only request is a couple of test cases to make sure this is as general as it sounds like it should be. Excellent work.

File: src/relax/transform/inline_functions.cc (hunk 2 above)
Comment (Lunderberg) [reply src/relax/transform/inline_functions.cc:127]: Thank you.  I may have gone a bit overboard with this comment, but I wanted to make sure that it was "clever robust" and not "clever unmaintainable".  I'm also really liking the use of CanonicalizeBindings as a general utility, since the rules about `DataflowVar` have too many edge cases to duplicate the logic everywhere.

File: tests/python/relax/test_inline_functions.py (hunk 4 above)
Comment (Lunderberg) [reply tests/python/relax/test_inline_functions.py:360]: Good catch.  I have test cases for theses two cases for the `InlinePrivateFunctions` transform, but not for the explicit user-driven calls.

Comment (Lunderberg) [reply tests/python/relax/test_inline_functions.py:360]: Good catch, thank you!  The inlining within an inlined function should already be tested in `test_inline_nested_subroutine_calls`.  The second case I had missed, and have now added `test_error_when_inlining_mutually_recursive_functions`.

File: src/relax/transform/inline_functions.cc (hunk 2 above)
Comment (slyubomirsky) [reply src/relax/transform/inline_functions.cc:127]: Yes, exactly.

File: tests/python/relax/test_inline_functions.py (hunk 3 above)
Comment (slyubomirsky) [reply tests/python/relax/test_inline_functions.py:173]: Ah, very nice.

**Review by slyubomirsky — APPROVED** [review]
Thank you for responding to my feedback. This is a very good change to have.
